src/ControlSystem/TimescaleTuner.cpp

```diff
@@ -0,0 +1,126 @@
+  if (min_timescale_ <= 0.0) {
+    ERROR("The specified minimum timescale " << min_timescale_
+                                             << " must be > 0");
+  }
+  if (max_timescale_ <= min_timescale_) {
+    ERROR("The maximum timescale "
+          << max_timescale_
+          << " must be > than the specified minimum timescale "
+          << min_timescale_);
+  }
+  if (min_threshold_ <= 0.0) {
+    ERROR("The specified minimum timescale " << min_timescale_
+                                             << " must be > 0");
+  }
+  if (max_threshold_ <= min_threshold_) {
+    ERROR("The maximum threshold "
+          << max_threshold_
+          << " must be > than the specified minimum threshold "
+          << min_threshold_);
+  }
+}
+
+void TimescaleTuner::suggested_timescale_if_in_allowable_range(
+    double suggested_timescale) noexcept {
+  for (auto& t_scale : timescale_) {
+    if (suggested_timescale > max_timescale_) {
+      t_scale = max_timescale_;
+    } else if (suggested_timescale < min_timescale_) {
+      t_scale = min_timescale_;
+    } else {
+      t_scale = suggested_timescale;
+    }
+  }
+}
+
+void TimescaleTuner::update_timescale(
+    const std::array<DataVector, 3>& qs) noexcept {
+  // above, Q is an array of length 3: \int Q, Q, dQ/dt
+  // each of these 3 is a DataVector w/ length = the number of components
+
+  if (qs[0].size() != timescale_.size()) {
+    ERROR("The number of components in qs("
+          << qs[0].size() << ") does not match the number of timescales("
+          << timescale_.size() << ")");
+  }
+  const DataVector& q = gsl::at(qs, 1);
+  const DataVector& dtq = gsl::at(qs, 2);
+
+  for (size_t i = 0; i < q.size(); i++) {
+    bool increase_timescale = false;
+    bool decrease_timescale = false;
+
+    // check whether we need to decrease the timescale:
+    // if Q or dtQ are above the maximum tolerance
+    if (fabs(q[i]) > max_threshold_ ||
+        fabs(dtq[i] * timescale_[i]) > max_threshold_) {
+      // AND
+      // (1) Q and dtQ are the same sign: the error is growing
+      // or (2) if Q is not expected to drop to half of its current value in
+      // one timescale: not decreasing fast enough
+      if (dtq[i] * q[i] > 0.0 ||
+          fabs(dtq[i]) * timescale_[i] < 0.5 * fabs(q[i])) {
+        decrease_timescale = true;
+      }
+    }
+
+    // check whether we need to increase the timescale:
+    // if Q and dtQ are below the minimum required threshold
+    if (fabs(q[i]) < min_threshold_ &&
+        fabs(dtq[i] * timescale_[i]) < (min_threshold_ - fabs(q[i]))) {
+      increase_timescale = true;
+    }
+
+    if (increase_timescale) {
+      timescale_[i] *= increase_factor_;
+    } else if (decrease_timescale) {
+      timescale_[i] *= decrease_factor_;
+    }
+
+    // make sure the timescale has not increased(decreased) above(below) the
+    // maximum(minimum) value.
+    timescale_[i] =
+        (timescale_[i] > max_timescale_) ? max_timescale_ : timescale_[i];
```

Comment:
The ? syntax is unnecessarily opaque. Please use if/else for clarity, unless there is some speed penalty to that I don't know about.

Reply:
No, if/else is not what should be used here, a ternary is the right answer. You are assigning a result based on a conditional. While the underlying generated code is likely the same, the if/else approach is 5x as many lines.

Reply:
Actually, @nils, the specific section I allude to is a nice discussion called "how tricky is too tricky". It's a good take on general philosophy, separate from other C++ decisions in Numerical Recipes that we would not make.

Reply:
Yes, given how the authors actually developed the software distributed with the book, I wouldn't put any value on their insight into this :)

Furthermore, ternary operators are taught alongside other conditionals, e.g. http://www.learncpp.com/cpp-tutorial/34-sizeof-comma-and-conditional-operators/

Reply:
`timescale_[i] = std::min(timescale_[i], max_timescale_);`

Reply:
Nice one-liner for this, @wthrowe! I like this for being both clear and concise.